packages/hub/services/search-index.ts

```diff
@@ -143,6 +149,19 @@ class IndexerRun implements IndexerHandle {
     let queue = [...this.touched.keys()];
     for (let i = 0; i < queue.length; i += queryBatchSize) {
       let queryRefs = queue.slice(i, i + queryBatchSize);
+
+      // TODO For each set of invalidations you may have graphs of dependencies:
+      // e.g. the card that was changed in the index was a grandparent card,
+      // whose parent, and grandchild were invalidated. We need to compile the
+      // parent card before the grandchild, as when the grandchild is compiled,
+      // it will use the parent's compiled card from the index. if we don't
+      // compile the parent card first, then the grandchild card will not
+      // contain the updated source from the original invalidation. In order to
+      // accomplish this consider converting the invalidated cards into a
+      // consumption graph, and then traverse it in such a way that we compiled
+      // the consumed cards before the consumer cards. (this is a pattern we
+      // used in CatalogJS that might be wroth repurposing)
+
```

Comment:
@ef4 @mattmcmanus so right now we have some flaky tests because of this which comes down to the order of reindexing. This is what the tests looks like when they pass:
``` 
  hub/search-index starting to index realm https://cardstack.local/
  hub/search-index reindexing https://cardstack.local/example because [
  hub/search-index   'https://cardstack.local/post',
  hub/search-index   'https://cardstack.com/base/base',
  hub/search-index   'https://cardstack.com/base/string'
  hub/search-index ]
  hub/search-index reindexing https://cardstack.local/grandchild because [
  hub/search-index   'https://cardstack.local/example',
  hub/search-index   'https://cardstack.local/post',
  hub/search-index   'https://cardstack.com/base/base',
  hub/search-index   'https://cardstack.com/base/string'
  hub/search-index ]
  hub/search-index finished indexing realm https://cardstack.local/
    ✓ can invalidate a card via the update of a grandparent adoptsFrom card (176ms)
```
this is what they look like when they fail: 
```
  hub/search-index starting to index realm https://cardstack.local/
  hub/search-index reindexing https://cardstack.local/grandchild because [
  hub/search-index   'https://cardstack.local/example',
  hub/search-index   'https://cardstack.local/post',
  hub/search-index   'https://cardstack.com/base/base',
  hub/search-index   'https://cardstack.com/base/string'
  hub/search-index ]
  hub/search-index reindexing https://cardstack.local/example because [
  hub/search-index   'https://cardstack.local/post',
  hub/search-index   'https://cardstack.com/base/base',
  hub/search-index   'https://cardstack.com/base/string'
  hub/search-index ]
  hub/search-index finished indexing realm https://cardstack.local/
    1) can invalidate a card via the update of a grandparent adoptsFrom card
```
note that in the case where it passes, the parent card is reindexed before the grandchild card
in the case where it fails, the grandchild card is reindexed before the parent card
i think it comes down to the fact that the grandchild is looking at the index for the compiled card of it's parent
in the case we index the parent after the grandchild, the parent does not yet have the new field
and hence the granchild does not yet have the new field